core/capabilities/launcher.go

```diff
@@ -118,6 +118,74 @@ func NewLauncher(
 	}
 }
 
+// CRE-326: Maintain only necessary Don2Don connections:
+//   - Workflow DONs connect only to other DONs that have at least one remote capability
+//   - Capability DONs connect only to workflow DONs
+//
+// Returns boolean as:
+//   - true: filter out
+//   - false: keep
+func filterDon2Don(logger logger.Logger, isCapabilityDON bool, isWorkflowDON bool, d registrysyncer.DON) bool {
+	// Below logic is based on identification who is who using a workflow acceptance flag:
+	//  d.DON.AcceptsWorkflows == true
+	// We identify few cases from the perspective of the node:
+	if isCapabilityDON && isWorkflowDON {
+		// as both workflow & capability DON let's just connect to anything
+		return false
+	}
+	if !isCapabilityDON && !isWorkflowDON {
+		// as none of workflow & capability DON don't use bandwidth
+		return true
+	}
+	if isWorkflowDON && d.DON.AcceptsWorkflows {
+		logger.Debugw("filterDon2Don: as a workflow DON my peers should be only capability DONs - filtering out", "DON.ID", d.ID)
+		return true
+	}
+	if isCapabilityDON && !d.DON.AcceptsWorkflows {
+		logger.Debugw("filterDon2Don: as a capability DON my peers should only be workflow DONs - filtering out", "DON.ID", d.ID)
+		return true // filter out
+	}
+	return false // keep
+}
+
+func (w *launcher) identifyPeers(isCapabilityDON bool, isWorkflowDON bool, allDONIDs []registrysyncer.DonID, state *registrysyncer.LocalRegistry) map[ragetypes.PeerID]p2ptypes.StreamConfig {
+	allPeers := make(map[ragetypes.PeerID]p2ptypes.StreamConfig)
+	for _, id := range allDONIDs {
+		d := state.IDsToDONs[id]
```

Comment:
Maybe rename "d" -> "candidatePeerDON" ? Or something to indicate that this is a DON member that we are considering connecting to...